### PR TITLE
feat: add zoom transcription

### PR DESCRIPTION
**Pull request by nickoferrall**
This PR adds a [Recall](https://www.recall.ai/) bot to a Zoom call and then adds the meeting transcription to the discussion thread at the end of the meeting.

Demo: https://www.loom.com/share/8220745c5780434c99e3a02b13214185

I think it should fix https://github.com/ParabolInc/parabol/issues/7674 too. See comment in my review.

Things to improve:
- Meeting settings UI & UX. Can't edit or remove the link at the moment
- Fix: after ending a retro meeting and going back to the meeting settings, the same Zoom URL still shows up until you refresh the page
- Live streaming the transcription during the meeting rather than at the end of it
- Add the whole meeting transcription to a slideable right sidebar rather than the discussion thread
- Each Discuss Stage could have the transcription that happened during that Stage. We get the timestamps of each sentence from Recall (or other providers) and we know when a Discuss Stage starts and ends, so this should be possible
- We could match the user name from Recall with the user name from Parabol, so the UI would look more like a comment with the user's avatar
- More ideas [here](https://parabol.slack.com/archives/C042E8PGPD4/p1679942702021109)

This feature isn't a core focus of our Squad, so I won't create a milestone and GitHub issues yet. We may have a better feel for what this feature should be once we've played around with it in Parabol meetings.

### To test

- [ ] Add the `zoomTranscription` org feature flag
- [ ] Add the `RECALL_AI_KEY` env var to your `.env`, see [here](https://github.com/ParabolInc/action-devops/blob/db3e3ded5f1cddb6f8607da9e1e391a5d5f0eab0/environments/local#L78)
- [ ] Create a Zoom meeting
- [ ] Open the Parabol meeting settings and paste the Zoom meeting URL 
- [ ] Start the Parabol meeting and see that the Parabol Notetaker bot has joined the Zoom meeting
- [ ] End the meeting, open the new meeting modal, refresh the page, open the settings, and see that the Zoom meeting URL is empty
- [ ] Go to the ended meeting, click on the Transcription tab in the discussion thread and see the Zoom transcription

### CODE REVIEW

File: packages/server/graphql/mutations/endRetrospective.ts [hunk 1]
```diff
@@ -80,10 +93,20 @@ const finishRetroMeeting = async (meeting: MeetingRetrospective, context: GQLCon
             .default(0) as unknown as number,
           topicCount: reflectionGroupIds.length,
           reflectionCount: reflections.length,
+          summary,
```
Comment (nickoferrall) [inline packages/server/graphql/mutations/endRetrospective.ts:96]: I think this should fix: https://github.com/ParabolInc/parabol/issues/7674

Previously, we were updating the Rethinkdb meeting object both here and in `generateWholeMeetingSummary`. By doing it in one place, we'll hopefully fix that race condition

File: packages/integration-tests/tests/retrospective-demo/step4-discuss.test.ts [hunk 2]
```diff
@@ -137,23 +137,25 @@ test.describe('retrospective-demo / discuss page', () => {
...
-          timeout: 30_000
-        })
-      }
-
-      for await (const comment of comments || []) {
-        await expect(page.locator(`[data-cy=comment-wrapper] :text('${comment}')`)).toBeVisible({
-          timeout: 30_000
-        })
-      }
+      // if (!isMobile) {
```
Comment (nickoferrall) [inline packages/integration-tests/tests/retrospective-demo/step4-discuss.test.ts:140]: I'm not sure why these tests are failing as I can still see the emojis, tasks, and comments on desktop with and without the Zoom Transcription flag.

I can only think that adding the transcription tab to the discussion thread has thrown them off.

I also don't see any reference to the data-cy `task-wrapper` or `comment-wrapper` in the codebase.

If you have any ideas why these are failing, please let me know! If not, I'll create an issue to fix them.

Comment (igorlesnenko) [reply packages/integration-tests/tests/retrospective-demo/step4-discuss.test.ts:140]: Can't say with a quick glance without debugging. I think we can create an issue to fix if you consider it is safe.

**Review by igorlesnenko — APPROVED** [review]
Works! I think it is safe to merge considering future refactoring

File: packages/server/graphql/types/RetrospectiveMeetingSettings.ts [hunk 3]
```diff
@@ -115,6 +116,14 @@ const RetrospectiveMeetingSettings: GraphQLObjectType<any, GQLContext> = new Gra
...
         return connectionFromTemplateArray(publicTemplates, first, after)
       }
+    },
+    recallBotId: {
+      type: GraphQLID,
+      description: 'The botId for the recall bot in the video meeting'
+    },
+    videoMeetingURL: {
+      type: GraphQLString,
+      description: 'The Zoom meeting URL for the meeting'
```
[the remark this answers is not in the document]
Comment (nickoferrall) [reply packages/server/graphql/types/RetrospectiveMeetingSettings.ts:126]: I just tested it again but I didn't get that error in the console 🤔  For sure, happy to add validation later

File: packages/server/graphql/mutations/endRetrospective.ts [hunk 4]
```diff
@@ -80,10 +93,20 @@ const finishRetroMeeting = async (meeting: MeetingRetrospective, context: GQLCon
...
         },
         {nonAtomic: true}
       )
+      .run(),
+    r
+      .table('MeetingSettings')
+      .get(settingsId)
+      .update({
+        recallBotId: null,
+        videoMeetingURL: null
```
[the remark this answers is not in the document]
Comment (nickoferrall) [reply packages/server/graphql/mutations/endRetrospective.ts:108]: > Feels like this should be submitted directly to the meeting object

Potentially! But when the user is selecting their retro meeting, the meeting object doesn't yet exist. A few options I can see:

1. Create the meeting object when the user adds the Zoom link in the settings UI
2. Make the user enter the Zoom link after starting the meeting
3. Continue with the current approach where it's stored in the Settings and then removed after the meeting ends

File: packages/integration-tests/tests/retrospective-demo/step4-discuss.test.ts (hunk 2 above)
Comment (Dschoordsch) [reply packages/integration-tests/tests/retrospective-demo/step4-discuss.test.ts:140]: -1 going through the demo throws an error
```
TypeError
Cannot read properties of undefined (reading 'zoomTranscription')
Call Stack
 eval
  packages/client/components/RetroDiscussionThreadHeader.tsx:108:128
 _optionalChain
  packages/client/components/RetroDiscussionThreadHeader.tsx:43:15
 RetroDiscussionThreadHeader
  packages/client/components/RetroDiscussionThreadHeader.tsx:108:40
 renderWithHooks
  parabol-action/./node_modules/react-dom/cjs/react-dom.development.js:14985:18
 mountIndeterminateComponent
  parabol-action/./node_modules/react-dom/cjs/react-dom.development.js:17811:13
 beginWork
  parabol-action/./node_modules/react-dom/cjs/react-dom.development.js:19049:16
 HTMLUnknownElement.callCallback
  parabol-action/./node_modules/react-dom/cjs/react-dom.development.js:3945:14
 Object.invokeGuardedCallbackDev
  parabol-action/./node_modules/react-dom/cjs/react-dom.development.js:3994:16
 invokeGuardedCallback
  parabol-action/./node_modules/react-dom/cjs/react-dom.development.js:4056:31
 beginWork$1
  parabol-action/./node_modules/react-dom/cjs/react-dom.development.js:23959:7
```

File: packages/server/graphql/types/NewMeeting.ts [hunk 5]
```diff
@@ -201,6 +201,10 @@ export const newMeetingFields = () => ({
       const viewerId = getUserId(authToken)
       return isMeetingLocked(viewerId, teamId, endedAt, dataLoader)
     }
+  },
+  transcription: {
```
Comment (Dschoordsch) [inline packages/server/graphql/types/NewMeeting.ts:205]: -1 I think we could go with this for a prototype, however going forward it's probably much more useful if we keep the context and attach the transcription to the meeting phases/stages

**Review by Dschoordsch — CHANGES_REQUESTED** [review]
Only looked at it high level

File: packages/server/graphql/types/NewMeeting.ts (hunk 5 above)
Comment (nickoferrall) [reply packages/server/graphql/types/NewMeeting.ts:205]: Agreed! We could look at the timestamp we get from Recall and add it to the correct phase & stage

File: packages/integration-tests/tests/retrospective-demo/step4-discuss.test.ts (hunk 2 above)
Comment (nickoferrall) [reply packages/integration-tests/tests/retrospective-demo/step4-discuss.test.ts:140]: Classic demo! Thanks, fixed

File: packages/client/components/DiscussionThreadListEmptyState.tsx [hunk 6]
```diff
@@ -36,26 +47,126 @@ const Message = styled('div')({
...
   allowTasks: boolean
+  settingsRef?: DiscussionThreadListEmptyState_settings$key
+  showTranscription?: boolean
+}
+
+const getMessage = (
+  allowTasks: boolean,
+  hasVideoURL: boolean,
+  isReadOnly: boolean,
+  showTranscription: boolean
```
[the remark this answers is not in the document]
Comment (Dschoordsch) [reply packages/client/components/DiscussionThreadListEmptyState.tsx:91]: Totally agree, this is very hacky atm.

**Review by Dschoordsch — CHANGES_REQUESTED** [review]

**Comment by nickoferrall** [comment]
Hey @mattkrick, I can see that there have been a few changes related to the build and yarn.lock file lately. I was wondering if they might be related to while this PR is failing the build at all? It's an old PR. 

I can't see anything particularly meaningful in the build logs: https://github.com/ParabolInc/parabol/actions/runs/5048056122/jobs/9055756147?pr=7949

**Comment by mattkrick** [comment]
The bug i see is 

```
Error [ERR_REQUIRE_ESM]: require() of ES Module /home/runner/work/parabol/parabol/node_modules/swiper/swiper.esm.js from /home/runner/work/parabol/parabol/node_modules/adaptivecards/lib/carousel.js not supported.
Instead change the require of swiper.esm.js in /home/runner/work/parabol/parabol/node_modules/adaptivecards/lib/carousel.js to a dynamic import() which is available in all CommonJS modules.
```

you can probably fix this by adding `mainFields: ['main']` to the webpack config.

**Comment by nickoferrall** [comment]
Thanks for looking into it, @mattkrick. Rebasing did the trick 👍

File: packages/server/graphql/mutations/endRetrospective.ts (hunk 1 above)
Comment (nickoferrall) [reply packages/server/graphql/mutations/endRetrospective.ts:97]: Bug fix moved to this PR: https://github.com/ParabolInc/parabol/pull/8197

**Comment by nickoferrall** [comment]
Would love to get this PR reviewed, @Dschoordsch 🙏  merge conflicts are becoming time consuming

**Comment by Dschoordsch** [comment]
Looking at it right now

**Review by Dschoordsch — APPROVED** [review]
Fine for me for now for internal testing.

**Pull request merged**